contracts/tournament/Tournament.cairo

```diff
@@ -327,4 +430,119 @@ func _only_tournament_closed{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, r
+
+    let (ship_count) = playing_ships_count_.read()
+    if ship_count == 1:
+        # This is the end! We have a single winner!
+        return ()
+    end
+
+    _rec_start()
+    return ()
+end
+
+# Recursively plays all battles
+func _rec_play_current_round_battles{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(battle_ship_index : felt):
+    
+    let (end_reached : felt, next_battle_ship_index : felt) = _play_next_battle(battle_ship_index)
+    if end_reached == TRUE:
+        return ()
+    end
+    
+    _rec_play_current_round_battles(next_battle_ship_index)
+    return ()
+end
+
+# Play the next battle entirely
+func _play_next_battle{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(ship_index : felt) -> (end_reached : felt, next_battle_ship_index : felt):
+    alloc_locals
+    let (ships_len : felt, ships : felt*) = _build_battle_ship_array(ship_index)
+    if ships_len == 0:
+        return (TRUE, 0)
+    end
+
+    let (space_contract) = space_contract_address_.read()
+    let (rand_contract) = rand_contract_address_.read()
+    let (grid_size) = grid_size_.read()
+    let (turn_count) = turn_count_.read()
+    let (max_dust) = max_dust_.read()
+
+    ISpace.play_game(space_contract,
+            rand_contract, grid_size, turn_count, max_dust,
+            ships_len, ships)
+    
+    let (played_battle_count) = played_battle_count_.read()
+    played_battle_count_.write(played_battle_count + 1)
+
+    # TODO: get scores
+    # TODO: get winner
+    let winner = ships[0]
+    let (winner_ships_count) = winner_ships_count_.read()
+    winner_ships_.write(winner_ships_count, winner)
+    winner_ships_count_.write(winner_ships_count+1)
+
+    return (FALSE, ship_index + ships_len)
+end
+
+# Get the ships that will participate in the next battle
+func _build_battle_ship_array{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(ship_index : felt) -> (
+        ships_len : felt, ships : felt*):
+    alloc_locals
+
+    let (local ships : felt*) = alloc()
+
+    let (ships_len) = _rec_build_battle_ship_array(ship_index, 0, ships)
+
+    return (ships_len, ships)
+end
+
+func _rec_build_battle_ship_array{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+        ship_index : felt, ships_len : felt, ships : felt*) -> (len : felt):
+    let (ships_per_battle) = ships_per_battle_.read()
+    if ships_len == ships_per_battle:
+        return (ships_len)
+    end
+
+    let (ship_count) = playing_ships_count_.read()
+    if ship_index == ship_count:
+        return (ships_len)
+    end
+
+    let (ship_address : felt) = playing_ships_.read(ship_index)
+    assert_not_zero(ship_address)
+
+    assert ships[ships_len] = ship_address
+
+    return _rec_build_battle_ship_array(ship_index + 1, ships_len + 1, ships)
+end
+
+func _update_playing_ships{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}():
+    let (ship_count) = winner_ships_count_.read()
```

Comment:
Might be good to be more explicit about this variable name: `winning_ships_count` or `previous_round_winners_count`?

Reply:
is it better now?

Reply:
Sorry my comment wasn't clear, I was referring to the variable `ship_count`, it would be better named `winning_ship_count`